tests/devices/experimental/test_default_qubit_2.py

```diff
@@ -453,8 +468,79 @@ def test_tf_backprop(self, convert_to_hamiltonian):
+
+    def test_single_circuit(self):
+        """Tests a basic example with a single circuit."""
+        dev = DefaultQubit2()
+        x = np.array(np.pi / 7)
+        qs = qml.tape.QuantumScript([qml.RX(x, 0)], [qml.expval(qml.PauliZ(0))])
+        qs = validate_and_expand_adjoint(qs)
+        expected_grad = -qml.math.sin(x)
+        actual_grad = dev.compute_derivatives(qs, self.ec)
+        assert isinstance(actual_grad, np.ndarray)
+        assert actual_grad.shape == ()
+        assert np.isclose(actual_grad, expected_grad)
+
+        expected_val = qml.math.cos(x)
+        actual_val, actual_grad = dev.execute_and_compute_derivatives(qs, self.ec)
+        assert np.isclose(actual_val, expected_val)
+        assert np.isclose(actual_grad, expected_grad)
+
+    def test_list_with_single_circuit(self):
+        """Tests a basic example with a batch containing a single circuit."""
+        dev = DefaultQubit2()
+        x = np.array(np.pi / 7)
+        qs = qml.tape.QuantumScript([qml.RX(x, 0)], [qml.expval(qml.PauliZ(0))])
+        qs = validate_and_expand_adjoint(qs)
+        expected_grad = -qml.math.sin(x)
+        actual_grad = dev.compute_derivatives([qs], self.ec)
+        assert isinstance(actual_grad, tuple)
+        assert isinstance(actual_grad[0], np.ndarray)
+        assert np.isclose(actual_grad[0], expected_grad)
+
+        expected_val = qml.math.cos(x)
+        actual_val, actual_grad = dev.execute_and_compute_derivatives([qs], self.ec)
+        assert np.isclose(expected_val, actual_val[0])
+        assert np.isclose(expected_grad, actual_grad[0])
+
+    def test_many_tapes_many_results(self):
+        """Tests a basic example with a batch of circuits of varying return shapes."""
+        dev = DefaultQubit2()
+        x = np.array(np.pi / 7)
+        single_meas = qml.tape.QuantumScript([qml.RX(x, 0)], [qml.expval(qml.PauliZ(0))])
+        multi_meas = qml.tape.QuantumScript(
+            [qml.RY(x, 0)], [qml.expval(qml.PauliX(0)), qml.expval(qml.PauliZ(0))]
+        )
+        expected_grad = (-qml.math.sin(x), (qml.math.cos(x), -qml.math.sin(x)))
+        actual_grad = dev.compute_derivatives([single_meas, multi_meas], self.ec)
+        assert np.isclose(actual_grad[0], expected_grad[0])
+        assert isinstance(actual_grad[1], tuple)
+        assert qml.math.allclose(actual_grad[1], expected_grad[1])
+
+    def test_integration(self):
+        """Tests the expected workflow done by a calling method."""
+        dev = DefaultQubit2()
+        x = np.array(np.pi / 7)
+        expected_grad = (-qml.math.sin(x), (qml.math.cos(x), -qml.math.sin(x)))
+        single_meas = qml.tape.QuantumScript([qml.RX(x, 0)], [qml.expval(qml.PauliZ(0))])
+        multi_meas = qml.tape.QuantumScript(
+            [qml.RY(x, 0)], [qml.expval(qml.PauliX(0)), qml.expval(qml.PauliZ(0))]
+        )
+
+        circuits, fn = dev.preprocess([single_meas, multi_meas], self.ec)
+        actual_grad = fn(dev.compute_derivatives(circuits, self.ec))
```

Comment:
So right now `fn` is for post-procssing the results.  Any classical post-processing will occur after we have left the black box boundary.  Amintor was confused about this too, so maybe I should go back and improve the documentation.

Within `interfaces` we will register the output of `compute_derivatives` as the derivative of `circuits`.  The ML interface will then manage the derivatives of `fn` to get us to the derivatives of `single_meas` and `multi_meas`.

Reply:
not entirely sure what you're getting at. You mean to say that the post-processing function is expecting the actual results, and not the derivatives?

Reply:
Yep. Post-processing is for applying to the results. It does not apply to the derivatives.